### PR TITLE
add cost calculation to client

### PR DESCRIPTION
**Pull request by kevin666aa**
## Why are these changes needed?

Add cost calculation to client in v0.2

## Related issue number

#614 

## Checks

- [ ] I've included any doc changes needed for https://microsoft.github.io/autogen/. See https://microsoft.github.io/autogen/docs/Contribute#documentation to build and test documentation locally.
- [ ] I've added tests (if relevant) corresponding to the changes introduced in this PR.
- [ ] I've made sure all auto checks have passed.

### CODE REVIEW

File: autogen/oai/client.py [hunk 1]
```diff
@@ -261,10 +261,23 @@ def yes_or_no_filter(context, response):
                     # Return the response if it passes the filter or it is the last client
                     response.config_id = i
                     response.pass_filter = pass_filter
-                    # TODO: add response.cost
+                    response.cost = self.cost(response)
                     return response
                 continue  # filter is not passed; try the next config
 
+    def cost(self, response):
```
Comment (joshkyh) [inline autogen/oai/client.py:268]:
```suggestion
    def cost(self, response: Union[ChatCompletion, Completion]) -> float:
```

File: autogen/oai/client.py [hunk 2]
```diff
@@ -261,10 +261,23 @@ def yes_or_no_filter(context, response):
...
+        """Calculate the cost of the response."""
+        model = response.model
+        if model not in oai_price1k:
+            return 0
+
+        n_input_tokens = response.usage.prompt_tokens
+        n_output_tokens = response.usage.completion_tokens
+        tmp_price1K = oai_price1k[model]
+        if isinstance(tmp_price1K, tuple):
```
Comment (joshkyh) [inline autogen/oai/client.py:277]:
```suggestion
        # First value is input token rate, second value is output token rate
        if isinstance(tmp_price1K, tuple):
```

File: autogen/oai/client.py [hunk 3]
```diff
@@ -261,10 +261,23 @@ def yes_or_no_filter(context, response):
...
-                    # TODO: add response.cost
+                    response.cost = self.cost(response)
                     return response
                 continue  # filter is not passed; try the next config
 
+    def cost(self, response):
+        """Calculate the cost of the response."""
+        model = response.model
+        if model not in oai_price1k:
+            return 0
```
Comment (joshkyh) [inline autogen/oai/client.py:272]:
```suggestion
            return 0 #TODO: When working on logging in the future, INFO logging that model is not found in price lookup.
```

**Review by joshkyh — COMMENTED** [review]
Great work! Quite a few people are asking for cost reports, it's great to have this feature (back)!

File: test/oai/test_client.py
[the remark this answers is not in the document]
Comment (kevin666aa) [reply test/oai/test_client.py:1]: Thank you for these suggestions, they are very useful!
For 1, that is not what I am thinking. No matter whether the cache_seed is set or not, we should let user know what cost of that creation is, even if it is from cache. Maybe we can inform the user about the actual cost and presumably cost (without cache) for a run when adding the feature of printing the usage. 
For 2, it is good to have an extra model tested!

**Review by qingyun-wu — COMMENTED** [review]
The code LGTM. Just one suggestion about documentation: We should make this feature discoverable with proper documentation or demonstration in a notebook.

**Comment by kevin666aa** [comment]
> The code LGTM. Just one suggestion about documentation: We should make this feature discoverable with proper documentation or demonstration in a notebook.

Got it. I just added the print cost in the documentation. I will add a notebook about cost calculation in the next PR when implementing cost summary of a client instance.

**Comment by afourney** [comment]
> The code LGTM. Just one suggestion about documentation: We should make this feature discoverable with proper documentation or demonstration in a notebook.

I'll add it to the Testbed as well (perhaps in another PR)

**Review by sonichi — APPROVED** [review]

**Pull request merged**